chia/wallet/coin_selection.py

```diff
@@ -0,0 +1,53 @@
+import random
+from typing import Set, Optional, Tuple, List
+from chia.types.blockchain_format.coin import Coin
+from chia.util.ints import uint64
+from chia.wallet.wallet_coin_record import WalletCoinRecord
+
+
+# we use this to check if one of the coins exactly matches the target.
+def check_for_exact_match(coin_records: List[WalletCoinRecord], target: int) -> Optional[WalletCoinRecord]:
+    for coinrecord in coin_records:
+        if coinrecord.coin.amount == target:
+            return coinrecord
+    return None
+
+
+# we use this to find the individual coin that is closest to the target amount.
+def find_smallest_coin(coin_records: List[WalletCoinRecord], target: int) -> Optional[WalletCoinRecord]:
+    smallest_value = float("inf")  # smallest coins value
+    smallest_coin: Optional[WalletCoinRecord] = None
+    for coinrecord in coin_records:
+        if target < coinrecord.coin.amount < smallest_value:
+            # try to find a coin that is as close as possible to the amount.
+            smallest_value = coinrecord.coin.amount
+            smallest_coin = coinrecord
+    return smallest_coin
+
+
+# we use this to find the smallest set of coins.

```

Comment:
to find the set of coins which have total value closest to the target, but at least the target.

Reply:
We should also add a link to the algorithm paper or website.

Reply:
Also, the input coins have to be sorted by descending value according to the paper, we should point this out in the comment. We should either check that the ordering is correct and throw if it's not, OR sort the smaller coins after receiving them.

Reply:
Fixed